stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/service/sync/PushMessageSyncHandler.kt

```diff
@@ -0,0 +1,171 @@
+    private val firebaseMessageParser: FirebaseMessageParser by lazy { FirebaseMessageParserImpl(notificationConfig) }
+
+    /**
+     * Should be called from [FirebaseMessagingService.onNewToken] function.
+     * It registers the Firebase token at Stream backend service. It's required by the backend to send
+     * push notifications via Firebase.
+     *
+     * @param token the token delivered by Firebase service to [com.google.firebase.messaging.FirebaseMessagingService]
+     * registered at your app.
+     */
+    public fun onNewToken(token: String) {
+        if (ChatClient.isInitialized) {
+            ChatClient.instance().onNewTokenReceived(token)
+        } else {
+            val syncConfig = syncModule.encryptedBackgroundSyncConfigStore.get()
+            syncConfig?.let {
+                val config = it
+                val user = User(id = config.userId)
+                val userToken = config.userToken
+                GlobalScope.launch(DispatcherProvider.IO) {
+                    val client = initClient(service, user, userToken, config.apiKey)
+                    client.onNewTokenReceived(token)
+                }
+            }
+        }
+    }
+
+    /**
+     * Should be called from [FirebaseMessagingService.onMessageReceived] function. It parses the [message],
+     * and starts the background sync operation proceeded with notification displaying.
+     * In case the [message] is not delivered from the Stream push notification provider no operation is executed.
+     *
+     * @param message the [RemoteMessage] delivered to [com.google.firebase.messaging.FirebaseMessagingService]
+     * registered at your app.
+     */
+    public fun onMessageReceived(message: RemoteMessage) {
+        if (!isStreamMessage(message)) {
+            return
+        }
+        createSyncNotificationChannel()
+        showForegroundNotification(notificationConfig.smallIcon)
+
+        GlobalScope.launch(DispatcherProvider.IO) {
+            val cid: String = firebaseMessageParser.parse(message).let { "${it.channelType}:${it.channelId}" }
+            if (ChatDomain.isInitialized && ChatClient.isInitialized) {
+                logger.logD("Starting the sync")
+                performSync(ChatDomain.instance(), cid, ChatClient.instance(), message)
+            } else {
+                val syncConfig = syncModule.encryptedBackgroundSyncConfigStore.get()
+                syncConfig?.let {
+                    val config = it
+                    val user = User(id = config.userId)
+                    val token = config.userToken
+                    val client = initClient(service, user, token, config.apiKey)
+                    val domain = initDomain(user, client)
+                    logger.logD("Starting the sync, config: $syncConfig")
+                    performSync(domain, cid, client, message)
+                }
+            }
+        }
+        service.stopForeground(true)
```

Comment:
This `stopForeground` call used to be in the body of the coroutine launched above, so it would run after sync has completed. Now we `launch` a coroutine, and don't wait for its completion, just immediately stop the service.